set-balance.md

```diff
@@ -168,6 +192,346 @@ A `Result` is considered an `Action`.
+         <k> repatriate_reserved(SLASHED, BENEFICIARY, AMOUNT)
+          => set_free_balance(BENEFICIARY, BENEFICIARY_FREE_BALANCE +Int minInt(SLASHED_RESERVED_BALANCE, AMOUNT))
+          ~> set_reserved_balance(SLASHED, SLASHED_RESERVED_BALANCE -Int minInt(SLASHED_RESERVED_BALANCE, AMOUNT))
+         ...
+         </k>
+         <accounts>
+           <account>
+             <accountID> SLASHED </accountID>
+             <reservedBalance> SLASHED_RESERVED_BALANCE </reservedBalance>
+             ...
+           </account>
+           <account>
+             <accountID> BENEFICIARY </accountID>
+             <reservedBalance> BENEFICIARY_RESERVED_BALANCE </reservedBalance>
+             <freeBalance> BENEFICIARY_FREE_BALANCE </freeBalance>
+             ...
+           </account>
+         </accounts>
+      requires BENEFICIARY_FREE_BALANCE +Int BENEFICIARY_RESERVED_BALANCE >Int 0
+```
+
+### Slashing
+
+Used to punish a node for violating the protocol.
+
+```k
+    syntax Action ::= "slash" "(" AccountId "," Int ")"
+ // ---------------------------------------------------
+    rule [slash]:
+         <k> slash(ACCOUNT, AMOUNT) => set_free_balance(ACCOUNT, FREE_BALANCE -Int AMOUNT) ... </k>
+         <accounts>
+           <account>
+             <accountID> ACCOUNT </accountID>
+             <freeBalance> FREE_BALANCE </freeBalance>
+             ...
+           </account>
+         </accounts>
+         <totalIssuance> TOTAL_ISSUANCE => TOTAL_ISSUANCE -Int AMOUNT </totalIssuance>
+      requires FREE_BALANCE >=Int AMOUNT
+
+    rule [slash-empty-free]:
+         <k> slash(ACCOUNT, AMOUNT)
+          => set_free_balance(ACCOUNT, 0)
+          ~> slash_reserved(ACCOUNT, AMOUNT -Int FREE_BALANCE)
+         ...
+         </k>
+         <accounts>
+           <account>
+             <accountID> ACCOUNT </accountID>
+             <freeBalance> FREE_BALANCE </freeBalance>
+             ...
+           </account>
+         </accounts>
+         <totalIssuance> TOTAL_ISSUANCE => TOTAL_ISSUANCE -Int FREE_BALANCE </totalIssuance>
+      requires FREE_BALANCE <Int AMOUNT
+```
+
+Reservation and unreservation of balances
+=========================================
+
+Used to move balance from free to reserved and visa versa.
+
+```k
+    syntax Action ::= reserve ( AccountId , Int )
+ // ---------------------------------------------
+    rule [reserve]:
+         <k> reserve(ACCOUNT, AMOUNT)
+          => set_reserved_balance(ACCOUNT, FREE_BALANCE +Int AMOUNT)
+          ~> set_free_balance(ACCOUNT, FREE_BALANCE -Int AMOUNT)
+         ...
+         </k>
+         <accounts>
+           <account>
+             <accountID> ACCOUNT </accountID>
+             <freeBalance> FREE_BALANCE </freeBalance>
+             <reservedBalance> RESERVED_BALANCE </reservedBalance>
+             ...
+           </account>
+         </accounts>
+      requires FREE_BALANCE >=Int AMOUNT
+       andBool ensure_can_withdraw(ACCOUNT, Reserve, FREE_BALANCE -Int AMOUNT)
```

Comment:
any checks against `EXISTENTIAL_DEPOSIT`?

Reply:
Oh nvm, `set*balance` will do that.

Reply:
See above